vmm/src/vm.rs

```diff
@@ -2098,6 +2101,7 @@ impl Vm {
             &self.cpu_manager,
             &self.memory_manager,
             &self.numa_nodes,
+            tpm_enabled,
```

Comment:
```suggestion
            self.config.lock().unwrap().tpm.is_some(),
```

Reply:
I couldn't make this change because it results in a cloud-hypervisor being stuck while trying to acquire a Mutex.

Below is the relevant trace from rust-gdb:

```
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x000055f453d75e85 in std::sys::unix::futex::futex_wait () at library/std/src/sys/unix/futex.rs:62
#2  0x000055f45313e308 in std::sys::unix::locks::futex::Mutex::lock_contended () at library/std/src/sys/unix/locks/futex.rs:63
#3  0x000055f453644f60 in std::sys::unix::locks::futex::Mutex::lock (self=0x55f455b37810)
    at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys/unix/locks/futex.rs:35
#4  0x000055f45362ecda in std::sys_common::mutex::MovableMutex::raw_lock (self=0x55f455b37810)
    at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys_common/mutex.rs:75
#5  0x000055f453341cf5 in std::sync::mutex::Mutex<vmm::vm_config::VmConfig>::lock<vmm::vm_config::VmConfig> (self=0x55f455b37810)
    at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sync/mutex.rs:266
#6  0x000055f45359530e in vmm::device_manager::{impl#5}::append_aml_bytes (self=0x7f24080027e8, bytes=0x7f240e170f18) at vmm/src/device_manager.rs:4294
#7  0x000055f4535ccdc9 in vmm::acpi::create_dsdt_table (device_manager=0x7f240e17ba80, cpu_manager=0x7f240e17bab8, memory_manager=0x7f240e17bac0) at vmm/src/acpi.rs:183
#8  0x000055f4535cec9f in vmm::acpi::create_acpi_tables (guest_mem=0x7f2408002dc0, device_manager=0x7f240e17ba80, cpu_manager=0x7f240e17bab8,
    memory_manager=0x7f240e17bac0, numa_nodes=0x7f240e17bb10, tpm_enabled=false) at vmm/src/acpi.rs:637
#9  0x000055f453385989 in vmm::vm::Vm::create_acpi_tables (self=0x7f240e17ba68) at vmm/src/vm.rs:2095
#10 0x000055f453385e7a in vmm::vm::Vm::boot (self=0x7f240e17ba68) at vmm/src/vm.rs:2135
#11 0x000055f4531d01bc in vmm::Vmm::vm_boot (self=0x7f240e17ba50) at vmm/src/lib.rs:572
....
```
I expected the parameter to be resolved to a bool value and the lock on `self.config` to be released before `acpi::create_acpi_tables` is invoked. But above invocation results in cloud-hypervisor being stuck with the above stack trace. 

Please let me know if there is a better way to address this issue.

Reply:
It's a deadlock. You were correct to save it to a temporary value first.

Reply:
Sorry for asking something wrong :-/

Reply:
@rbradford why does this result in a deadlock? Doesn't Rust evaluate the parameter before it is passed to the function? I couldn't find any relevant information while googling about this behavior.

Reply:
I know about this form of re-entrancy deadlock from experience. I also couldn't find any documentation on this.